### ISSUE TITLE
Prefetch_related has strange behavior

### ISSUE THREAD

**Issue by EliotBerriot** [django-polymorphic/django-polymorphic#68]
Hi,

I use django-polymorphic for a personal, non-critical project, in which items (Storable) can be stored in workspaces (Workspace). 

This project relies heavily on concrete inheritance, and my models all inherit from a base model (using Polymorphic).

Everything works fine but, as always with concrete inheritance, database load is quite high, especially when I want to display a list of Storable, along with their workspaces.
In order to reduce the number of queries, I'm trying to use `prefetch_related('workspaces')` on Storable queryset. However, while it reduces indeed the number of queries, there is a bug when come the time to display items.

I've set up a test project with the following code.
# Models

```
from django.db import models
from polymorphic import PolymorphicModel


class Base(PolymorphicModel):
    name = models.CharField('name', max_length=255, default=None, blank=True, null=True)


class Storable(Base):
    pass


class Workspace(Base):
    items = models.ManyToManyField(Storable, related_name="workspaces", null=True, blank=True, default = None)
```
# View

```
from django.views.generic.list import ListView
from test_prefetch_related.models import Storable


class TestPrefetchRelated(ListView):
    model = Storable
    template_name = "test_prefetch_related/test_prefetch_related.html"
    paginate_by = 10

    def get_queryset(self):
        return self.model.objects.all().prefetch_related('workspaces')
```
# Template

```
<!DOCTYPE html>
<html>
<head>
    <title></title>
</head>
<body>
    {% for item in object_list %}
        <h1>{{ item.name }}</h1>
        <ul>
            {% for w in item.workspaces.all %}
                <li>{{ w.id }} - {{ w.name }}</li>
            {% empty %}
                None
            {% endfor %}
        </ul>
    {% endfor %}
</body>
</html>
```
# Database population

```
from test_prefetch_related.models import Storable, Workspace

# here is the code I used to populate the database with some data

# create workspaces
w1 = Workspace(name="w1")
w1.save()

w2 = Workspace(name="w2")
w2.save()

w3 = Workspace(name="w3")
w3.save()

# create storables
s1 = Storable(name="s1")
s1.save()
s1.workspaces.add(w1)
s1.save()

s2 = Storable(name="s2")
s2.save()
s2.workspaces.add(w2)
s2.save()

s3 = Storable(name="s3")
s3.save()
s3.workspaces.add(w1, w2, w3)  # this storable get all workpaces
s3.save()
```
# Html output

```
s1

    1 - w1
    1 - w1

s2

    2 - w2
    2 - w2

s3

    3 - w3
```
# The bug

It's hard to explain (excuse my english ;), but as you can see in output, workspaces are not displayed properly, next to each Storable. Intead, they are grouped on the first item in the queryset with which they are linked.

The expected output would be :

```
s1

    1 - w1

s2

    2 - w2

s3     

    1 - w1
    2 - w2
    3 - w3
```

I'm almost sure it is related to polymorphic because the following code using standard django inheritance works as expected.

In polymorphic docs, I did not found any mention of prefetch_related(). Is this behaviour one of the caveats of using concrete inheritance ? Is there anything else I could do to achieve the same result ?

You can download the code I used, if you want to check by yourself : http://seafile.eliotberriot.com/d/aca8f2399a/ (click on ZIP button in top right corner).

Thank you for your help and for this project, apart from that, I'm really happy with it !

**Comment by Franr** [comment]
Does anyone found a solution for this? If you use the non_polymorphic's approach, you're losing all the specific attributes and methods of each subclass.

**Comment by BertrandBordage** [comment]
The solution I found to that is to stop using django_polymorphic ;)

**Comment by NotSqrt** [comment]
Simple code to reproduce the problem:

```python
class ModelA(PolymorphicModel):
    pass

class ModelB(models.Model):
    links = models.ManyToManyField(ModelA)

b1 = ModelB.objects.create()
b2 = ModelB.objects.create()
a = b1.links.create()
b2.links.add(a)

for obj in ModelB.objects.prefetch_related('links'):
    print(obj, obj.links.all())
# returns 1 ModelB instance linked to 2 modelA, and the other with 0, instead of 1 for each

```

**Comment by vdboor** [comment]
Thanks @NotSqrt - that clears things up for me!
Looks like this needs a good unittest + perhaps a hack based on `_next_is_sticky()` to fix this?

**Comment by NotSqrt** [comment]
@vdboor IIRC, it's due to the ``base_result_objects_by_id`` cache in https://github.com/chrisglass/django_polymorphic/blob/master/polymorphic/query.py#L193 that will use the same instance of ``ModelA``, instead of duplicating the instance so that each ``ModelB`` can have its own ``ModelA`` version

**Comment by DylanLukes** [comment]
Just ran into this issue. Glad to see there are workarounds. Is there any work on this currently?

**Comment by domface** [comment]
+1

**Comment by mbarszcz** [comment]
+1

**Comment by Sajam** [comment]
+1

**Comment by Franr** [comment]
+2

**Comment by ish-vasa** [comment]
+1

**Comment by slamice** [comment]
+1

**Comment by jarrodw** [comment]
+1

**Comment by EliotBerriot** [comment]
Hi everyone, I started to work yesterday on this issue. I opened a pull request regarding that, this PR only contains a test reproducing the issue at the moment (https://github.com/django-polymorphic/django-polymorphic/issues/250).  

If you want to help, I think the discussion should continue on this PR, since we know the problem and only need a fix right now. Any help would much appreciated : I started to track down the issue, and it involves hardcore django's ORM code. The more eyes we get, the faster we can find a fix.
